2_process/src/lake_sequences_mntoha.py

```diff
@@ -0,0 +1,228 @@
+        lake_sequences = np.empty((0))
+    else:
+        # Read metadata with elevation
+        lake_metadata_augmented = pd.read_csv(metadata_augmented_file)
+        # NOTE: If there are duplicate sites in lake_metadata_augmented,
+        # only the first will be used.
+        # If there are no matches, an IndexError will be raised.
+        lake = lake_metadata_augmented[lake_metadata_augmented.site_id==site_id].iloc[0,:]
+
+        # Create one large DataFrame of temperatures at all days and all discretized depths
+        obs_full = all_dates_depths(lake_obs_interpolated,
+                                    depths,
+                                    pad_before_days=spinup_time,
+                                    min_days=sequence_length)
+        # Now obs_full has dimensions of len(date_range) by len(depths)
+        # Values are NaN wherever/whenever there are no observations
+
+        # There are some NaN values in the clarity and ice flag data.
+        # There are multiple ways to deal with NaN values in inputs:
+        # interpolate to fill them, skip over them, or remove sequences
+        # containing NaNs entirely. At this point, allow NaN values into the
+        # sites. Later code will handle NaN values.
+
+        # Join drivers by date
+        drivers = pd.read_csv(drivers_file, parse_dates=['time'], index_col='time')
+        obs_full = obs_full.join(drivers, how='left')
+
+        # Join clarity by date
+        clarity = pd.read_csv(clarity_file, parse_dates=['date'], index_col='date')
+        obs_full = obs_full.join(clarity, how='left')
+
+        # Join ice flags by date
+        ice_flags = pd.read_csv(ice_flags_file, parse_dates=['date'], index_col='date')
+        obs_full = obs_full.join(ice_flags, how='left')
+
+        # Get attributes
+        obs_full['area'] = lake['area']
+        obs_full['lon'] = lake['centroid_lon']
+        obs_full['lat'] = lake['centroid_lat']
+        obs_full['elevation'] = lake['elevation']
+        # Now obs_full is one long timeseries with depth-specific observations,
+        # drivers, clarity, ice_flags, and attributes as columns
+
+        # Split into sequences
+
+        # Let's make some rules for sequences:
+        # 1. Sequences are sequence_length days long
+        # 2. Each next sequence starts sequence_offset days later than the previous
+        # 3. Discard sequences that lack observations
+        # 4. We know going in that the first and last sequences have observations
+
+        # Convert dataframe to numpy array
+        obs_full_array = obs_full.to_numpy(dtype=np.float32)
+        # Create a strided view into obs_full_array
+        # This creates a numpy object with sequences of length sequence_length
+        # and offsets of sequence_offset, without making an in-memory copy of the array
+        all_sequences = np.lib.stride_tricks.sliding_window_view(
+            obs_full_array, sequence_length, axis=0
+        )[::sequence_offset]
```

Comment:
Okay @AndyMcAliley I just want to confirm that I understand what is happening here.

Here's my take:

The first bit `np.lib.stride_tricks.sliding_window_view(obs_full_array, sequence_length, axis=0)` is generating a sliding window 'view' into the `obs_full_array`. `obs_full_array` is shape (`date_range`, #depths + #features). Because you are specifying an integer (`sequence_length`) as the `window_shape` argument to `sliding_window_view()` and `axis=0`, the sliding window is shape `(sequence_length, )` and slides vertically across the rows (dates), generating a window that starts with the first date, then another that starts with the second date, and so on. The shape of the 'view' returned by that function call is (#windows, #depths + #features, `sequence_length`).

The last bit `[::sequence_offset]` pulls elements from the 3D returned 'view' array along the first dimension (#windows) with a step size of `sequence_offset`. So the first window -- with shape (#depths + #features, `sequence_length`) -- is pulled and then every nth window (where n = `sequence_offset`) is pulled until the end of the #windows dimension is reached.

In the end, we have an array with shape (# sequences, #depths + # features, `sequence_length`), where each sequence is offset from the previous sequence by `sequence_offset` days).

If that is a correct interpretation, I wonder if you might change that last bit to read `[::sequence_offset,:,:]`. I know it's not technically  necessary, but might make it clearer that the step indexing is only being applied to the first dimension of the returned array (#windows), and that the other dimensions (#depths + #features and `sequence_length`) are being pulled in full.

Reply:
Yes, that's it! Sure, I can add the other dimensions in the slice operation.